src/connection.cpp

```diff
@@ -43,20 +45,39 @@ void connection::stop() {
 void connection::handle_read(const boost::system::error_code& e,
     std::size_t bytes_transferred) {
   if (!e) {
+    boost::tribool result;
+    boost::tie(result, boost::tuples::ignore) = request_parser_.parse(request_, buffer_.data(), buffer_.data() + bytes_transferred);
 
-    if (buffer_[0] != '\0') {
-      request_handler_.handle_request(buffer_, reply_);
+    if (result) {
+
+      // TODO: request type isn't being set correctly, need to set up logging to debug
+
+      // Use parsed request to determine whether it wants an echo or a file
+      // if (request_.request_type == request::static_file) {
+      //   request_handler_static_.handle_request(request_, reply_);
+      // }
+      // else {
+      //   request_handler_echo_.handle_request(request_, reply_);
+      // }
+      request_handler_echo_.handle_request(request_, reply_);
       boost::asio::async_write(socket_, reply_.to_buffers(),
-          boost::bind(&connection::handle_write, shared_from_this(),
-            boost::asio::placeholders::error));
+        boost::bind(&connection::handle_write, shared_from_this(),
+          boost::asio::placeholders::error));
     }
-    else {
+    else if (!result) {
       reply_ = reply::stock_reply(reply::bad_request);
       boost::asio::async_write(socket_, reply_.to_buffers(),
-          boost::bind(&connection::handle_write, shared_from_this(),
-            boost::asio::placeholders::error));
+        boost::bind(&connection::handle_write, shared_from_this(),
+          boost::asio::placeholders::error));
+    }
+    else {
```

Comment:
Is this necessary? I believe if (result) and if (!result) covers all situations. Never hurts to be safe though.